File: C#/TheOpenMathLibrary.Engineering/OpticsAndPhotonics.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace TheOpenMathLibrary.Engineering
{
    /// <summary>
    /// This class contains functions for calculating various properties of light and electromagnetic radiation. It includes functions for calculating magnification, image height, image distance, poynting vector, poynting flux, RMS electric field, radiation momentum, radiant intensity, and radiosity. These functions can be used in physics and engineering applications to analyze and design optical systems.
    /// </summary>
    public class OpticsAndPhotonics
    {
        /// <summary>
        /// magnification function
        /// </summary>
        /// <param name="imageDistance"></param>
        /// <param name="objectDistance"></param>
        /// <returns></returns>
        public static double Magnification(double imageDistance, double objectDistance)
        {
            double magnification = 0;
            magnification = imageDistance / objectDistance;
            return magnification;
        }

        /// <summary>
        /// image height function
        /// </summary>
        /// <param name="magnification"></param>
        /// <param name="objectHeight"></param>
        /// <returns></returns>
        public static double ImageHeight(double magnification, double objectHeight)
        {
            double imageHeight = 0;
            imageHeight = magnification * objectHeight;
            return imageHeight;
        }

        /// <summary>
        /// image distance function 
        /// </summary>
        /// <param name="focalLength"></param>
        /// <param name="objectDistance"></param>
        /// <returns></returns>
        public static double ImageDistance(double focalLength, double objectDistance)
        {
            double imageDistance = 0;
            imageDistance = (focalLength * objectDistance) / (objectDistance - focalLength);
            return imageDistance;
        }

        /// <summary>
        /// Physical optics poynting vector function 
        /// </summary>
        /// <param name="electricField"></param>
        /// <param name="magneticField"></param>
        /// <returns></returns>
        public static double PoyntingVector(double electricField, double magneticField)
        {
            double poyntingVector = 0;
            poyntingVector = electricField * magneticField;
            return poyntingVector;
        }

        /// <summary>
        /// poyting flux, EM field power flow function 
        /// </summary>
        /// <param name="electricField"></param>
        /// <param name="magneticField"></param>
        /// <returns></returns>
        public static double PoyntingFlux(double electricField, double magneticField)
        {
            double poyntingFlux = 0;
            poyntingFlux = electricField * magneticField;
            return poyntingFlux;
        }

        /// <summary>
        /// RMS electric field of Light function 
        /// </summary>
        /// <param name="electricField"></param>
        /// <returns></returns>
        public static double RMSElectricField(double electricField)
        {
            double rmsElectricField = 0;
            rmsElectricField = electricField / Math.Sqrt(2);
            return rmsElectricField;
        }

        /// <summary>
        /// Radiation momentum function 
        /// </summary>
        /// <param name="intensity"></param>
        /// <param name="speedOfLight"></param>
        /// <returns></returns>
        public static double RadiationMomentum(double intensity, double speedOfLight)
        {
            double radiationMomentum = 0;
            radiationMomentum = intensity / speedOfLight;
            return radiationMomentum;
        }

        /// <summary>
        /// radiant intensity function 
        /// </summary>
        /// <param name="power"></param>
        /// <param name="solidAngle"></param>
        /// <returns></returns>
        public static double RadiantIntensity(double power, double solidAngle)
        {
            double radiantIntensity = 0;
            radiantIntensity = power / solidAngle;
            return radiantIntensity;
        }

        /// <summary>
        /// Radiosity function 
        /// </summary>
        /// <param name="emissivity"></param>
        /// <param name="stefanBoltzmannConstant"></param>
        /// <param name="temperature"></param>
        /// <returns></returns>
        public static double Radiosity(double emissivity, double stefanBoltzmannConstant, double temperature)
        {
            double radiosity = 0;
            radiosity = emissivity * stefanBoltzmannConstant * Math.Pow(temperature, 4);
            return radiosity;
        }

        /// <summary>
        /// Spectral radiance function 
        /// </summary>
        /// <param name="radiance"></param>
        /// <param name="wavelength"></param>
        /// <returns></returns>
        public static double SpectralRadiance(double radiance, double wavelength)
        {
            double spectralRadiance = 0;
            spectralRadiance = radiance / wavelength;
            return spectralRadiance;
        }

        /// <summary>
        /// spectral irradiance function 
        /// </summary>
        /// <param name="irradiance"></param>
        /// <param name="wavelength"></param>
        /// <returns></returns>
        public static double SpectralIrradiance(double irradiance, double wavelength)
        {
            double spectralIrradiance = 0;
            spectralIrradiance = irradiance / wavelength;
            return spectralIrradiance;
        }

        /// <summary>
        /// Energy density in an EM wave 
        /// </summary>
        /// <param name="electricField"></param>
        /// <param name="magneticField"></param>
        /// <returns></returns>
        public static double EnergyDensity(double electricField, double magneticField)
        {
            double energyDensity = 0;
            energyDensity = 0.5 * (electricField * electricField + magneticField * magneticField);
            return energyDensity;
        }

        /// <summary>
        /// Kinetic and potential momenta of light 
        /// </summary>
        /// <param name="speedOfLight"></param>
        /// <param name="mass"></param>
        /// <returns></returns>
        public static double KineticMomentum(double speedOfLight, double mass)
        {
            double kineticMomentum = 0;
            kineticMomentum = speedOfLight * mass;
            return kineticMomentum;
        }

        /// <summary>
        /// Doppler effect for light 
        /// </summary>
        /// <param name="frequency"></param>
        /// <param name="speedOfLight"></param>
        /// <param name="observerSpeed"></param>
        /// <returns></returns>
        public static double DopplerEffect(double frequency, double speedOfLight, double observerSpeed)
        {
            double dopplerEffect = 0;
            dopplerEffect = frequency * (speedOfLight + observerSpeed) / speedOfLight;
            return dopplerEffect;
        }

        /// <summary>
        /// cherenkov radiation 
        /// </summary>
        /// <param name="speedOfLight"></param>
        /// <param name="refractiveIndex"></param>
        /// <returns></returns>
        public static double CherenkovRadiation(double speedOfLight, double refractiveIndex)
        {
            double cherenkovRadiation = 0;
            cherenkovRadiation = speedOfLight / refractiveIndex;
            return cherenkovRadiation;
        }

        /// <summary>
        /// EM wave components 
        /// </summary>
        /// <param name="electricField"></param>
        /// <param name="magneticField"></param>
        /// <returns></returns>
        public static double EMWaveComponent(double electricField, double magneticField)
        {
            double emWaveComponent = 0;
            emWaveComponent = Math.Sqrt(electricField * electricField + magneticField * magneticField);
            return emWaveComponent;
        }


        /// <summary>
        /// Critical angle function 
        /// </summary>
        /// <param name="refractiveIndex1"></param>
        /// <param name="refractiveIndex2"></param>
        /// <returns></returns>
        public static double CriticalAngle(double refractiveIndex1, double refractiveIndex2)
        {
            double criticalAngle = 0;
            criticalAngle = Math.Asin(refractiveIndex2 / refractiveIndex1);
            return criticalAngle;
        }

        /// <summary>
        /// Thin lens equation function 
        /// </summary>
        /// <param name="focalLength"></param>
        /// <param name="objectDistance"></param>
        /// <param name="imageDistance"></param>
        /// <returns></returns>
        public static double ThinLensEquation(double focalLength, double objectDistance, double imageDistance)
        {
            double thinLensEquation = 0;
            thinLensEquation = 1 / focalLength - 1 / objectDistance + 1 / imageDistance;
            return thinLensEquation;
        }

        /// <summary>
        /// Image distance in a plane mirror function 
        /// </summary>
        /// <param name="objectDistance"></param>
        /// <returns></returns>
        public static double ImageDistancePlaneMirror(double objectDistance)
        {
            double imageDistancePlaneMirror = 0;
            imageDistancePlaneMirror = -objectDistance;
            return imageDistancePlaneMirror;
        }

        /// <summary>
        /// Spherical mirror equation function 
        /// </summary>
        /// <param name="focalLength"></param>
        /// <param name="objectDistance"></param>
        /// <param name="imageDistance"></param>
        /// <returns></returns>
        public static double SphericalMirrorEquation(double focalLength, double objectDistance, double imageDistance)
        {
            double sphericalMirrorEquation = 0;
            sphericalMirrorEquation = 1 / focalLength - 1 / objectDistance + 1 / imageDistance;
            return sphericalMirrorEquation;
        }
    }
}

File: Go/TheOpenMathLibrary/Engineering/OpticsAndPhotonics.go
package engineering

import (
	"math"
)

// OpticsAndPhotonics contains functions for calculating various properties of light and electromagnetic radiation.
type OpticsAndPhotonics struct{}

// Magnification calculates the magnification given image distance and object distance.
func Magnification(imageDistance, objectDistance float64) float64 {
	return imageDistance / objectDistance
}

// ImageHeight calculates the image height given magnification and object height.
func ImageHeight(magnification, objectHeight float64) float64 {
	return magnification * objectHeight
}

// ImageDistance calculates the image distance given focal length and object distance.
func ImageDistance(focalLength, objectDistance float64) float64 {
	return (focalLength * objectDistance) / (objectDistance - focalLength)
}

// PoyntingVector calculates the Poynting vector given electric field and magnetic field.
func PoyntingVector(electricField, magneticField float64) float64 {
	return electricField * magneticField
}

// PoyntingFlux calculates the Poynting flux given electric field and magnetic field.
func PoyntingFlux(electricField, magneticField float64) float64 {
	return electricField * magneticField
}

// RMSElectricField calculates the RMS electric field given electric field.
func RMSElectricField(electricField float64) float64 {
	return electricField / math.Sqrt(2)
}

// RadiationMomentum calculates the radiation momentum given intensity and speed of light.
func RadiationMomentum(intensity, speedOfLight float64) float64 {
	return intensity / speedOfLight
}

// RadiantIntensity calculates the radiant intensity given power and solid angle.
func RadiantIntensity(power, solidAngle float64) float64 {
	return power / solidAngle
}

// Radiosity calculates the radiosity given emissivity, Stefan-Boltzmann constant, and temperature.
func Radiosity(emissivity, stefanBoltzmannConstant, temperature float64) float64 {
	return emissivity * stefanBoltzmannConstant * math.Pow(temperature, 4)
}

// SpectralRadiance calculates the spectral radiance given radiance and wavelength.
func SpectralRadiance(radiance, wavelength float64) float64 {
	return radiance / wavelength
}

// SpectralIrradiance calculates the spectral irradiance given irradiance and wavelength.
func SpectralIrradiance(irradiance, wavelength float64) float64 {
	return irradiance / wavelength
}

// EnergyDensity calculates the energy density in an EM wave given electric field and magnetic field.
func EnergyDensity(electricField, magneticField float64) float64 {
	return 0.5 * (electricField*electricField + magneticField*magneticField)
}

// KineticMomentum calculates the kinetic momentum of light given speed of light and mass.
func KineticMomentum(speedOfLight, mass float64) float64 {
	return speedOfLight * mass
}

// DopplerEffect calculates the Doppler effect for light given frequency, speed of light, and observer speed.
func DopplerEffect(frequency, speedOfLight, observerSpeed float64) float64 {
	return frequency * (speedOfLight + observerSpeed) / speedOfLight
}

// CherenkovRadiation calculates the Cherenkov radiation given speed of light and refractive index.
func CherenkovRadiation(speedOfLight, refractiveIndex float64) float64 {
	return speedOfLight / refractiveIndex
}

// EMWaveComponent calculates the EM wave components given electric field and magnetic field.
func EMWaveComponent(electricField, magneticField float64) float64 {
	return math.Sqrt(electricField*electricField + magneticField*magneticField)
}

// CriticalAngle calculates the critical angle given two refractive indices.
func CriticalAngle(refractiveIndex1, refractiveIndex2 float64) float64 {
	return math.Asin(refractiveIndex2 / refractiveIndex1)
}

// ThinLensEquation calculates the thin lens equation given focal length, object distance, and image distance.
func ThinLensEquation(focalLength, objectDistance, imageDistance float64) float64 {
	return 1/focalLength - 1/objectDistance + 1/imageDistance
}

// ImageDistancePlaneMirror calculates the image distance in a plane mirror given object distance.
func ImageDistancePlaneMirror(objectDistance float64) float64 {
	return -objectDistance
}

// SphericalMirrorEquation calculates the spherical mirror equation given focal length, object distance, and image distance.
func SphericalMirrorEquation(focalLength, objectDistance, imageDistance float64) float64 {
	return 1/focalLength - 1/objectDistance + 1/imageDistance
}
